network/netplan/netplan.go

```diff
@@ -0,0 +1,245 @@
+	// create a bridge
+	if np.Network.Bridges == nil {
+		np.Network.Bridges = make(map[string]Bridge)
+	}
+	np.Network.Bridges[bridgeName] = Bridge{
+		Interfaces: []string{deviceId},
+		Interface:  intf,
+	}
+	np.Network.Ethernets[deviceId] = ethernet
+	return nil
+}
+
+func Unmarshal(in []byte, out interface{}) (err error) {
+	return goyaml.UnmarshalStrict(in, out)
+}
+
+func Marshal(in interface{}) (out []byte, err error) {
+	return goyaml.Marshal(in)
+}
+
+// readYamlFile reads netplan yaml into existing netplan structure
+// TODO(wpk) 2017-06-14 When reading files sequentially netplan replaces single
+// keys with new values, we have to simulate this behaviour.
+func (np *Netplan) readYamlFile(path string) (err error) {
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	err = Unmarshal(contents, np)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ReadDirectory Reads the contents of a netplan directory and
+// returns complete config.
+func ReadDirectory(dirPath string) (np Netplan, err error) {
+	fileInfos, err := ioutil.ReadDir(dirPath)
+	if err != nil {
+		return np, err
+	}
+	np.sourceDirectory = dirPath
+	for _, fileInfo := range fileInfos {
+		if !fileInfo.IsDir() && strings.HasSuffix(fileInfo.Name(), ".yaml") {
+			np.sourceFiles = append(np.sourceFiles, fileInfo.Name())
+		}
+	}
+	for _, fileName := range np.sourceFiles {
+		err := np.readYamlFile(path.Join(np.sourceDirectory, fileName))
+		if err != nil {
+			return np, err
+		}
+	}
+	return np, nil
+}
+
+// MoveYamlsToBak moves source .yaml files in a directory to .yaml.backup.(timestamp), except
+func (np *Netplan) MoveYamlsToBak() (err error) {
+	if np.backedFiles != nil {
+		return errors.Errorf("Cannot backup netplan yamls twice")
+	}
+	suffix := fmt.Sprintf(".bak.%d", time.Now().Unix())
+	np.backedFiles = make(map[string]string)
+	for _, file := range np.sourceFiles {
+		newFilename := fmt.Sprintf("%s%s", file, suffix)
+		oldFile := path.Join(np.sourceDirectory, file)
+		newFile := path.Join(np.sourceDirectory, newFilename)
+		// If it fails we're already past the point of no return, we might log it
+		_ = os.Rename(oldFile, newFile)
+		np.backedFiles[oldFile] = newFile
+	}
+	return nil
+}
+
+// Write writes merged netplan yaml to file specified by path. If path is empty filename is autogenerated
+func (np *Netplan) Write(inPath string) (filePath string, err error) {
+	if np.writtenFile != "" {
+		return "", errors.Errorf("Cannot write the same netplan twice")
+	}
+	if inPath == "" {
+		i := 99
+		for ; i > 0; i-- {
+			filePath = path.Join(np.sourceDirectory, fmt.Sprintf("%0.2d-juju.yaml", i))
+			_, err = os.Stat(filePath)
+			if os.IsNotExist(err) {
+				break
+			}
+		}
+		if i == 0 {
+			return "", errors.Errorf("Can't generate a filename for netplan YAML")
+		}
+	} else {
+		filePath = inPath
+	}
+	out, err := Marshal(np)
+	if err != nil {
+		return "", err
+	}
+	err = ioutil.WriteFile(filePath, out, 0644)
```

Comment:
some care to write to a .tmp and then rename into place so we don't have a partially written .yaml file that ends up in place.